compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp

```diff
@@ -473,5 +475,122 @@ void populateRemoveDeadMemAllocPatterns(RewritePatternSet &patterns) {
   patterns.insert<RemoveDeadInterfaceBindings>(patterns.getContext());
 }
 
+void analyseAllocsForPacking(func::FuncOp funcOp, ArrayRef<Operation *> allocs,
+                             SmallVector<AliasGroup> &aliasGroups) {
+  // Represent of a group of allocations with overlapping liverange and the
+  // liveness of the overall group.
+  struct AllocGroup {
+    SmallVector<Operation *> allocs;
+    // Keep track of every operation where any of the alloc in the group is
+    // live.
+    // Liveness is represent as a set of Operations where the alloc is alive.
+    // To make it merge liveranges and check if a given Operation interfers
+    // with the liverange we store it as a DesneSet.
+    llvm::DenseSet<Operation *> liveness;
+  };
+  Liveness liveness(funcOp);
+  SmallVector<AllocGroup> groups;
+  for (Operation *alloc : allocs) {
+    SmallVector<size_t> aliasGroups;
+    for (size_t i : llvm::seq<size_t>(0, groups.size())) {
+      AllocGroup &group = groups[i];
+      for (Operation *user : alloc->getUsers()) {
+        // Skip the whole analysis if any user is a subview.
+        // TODO: This could be extended if needed by recursively merging
+        // liveness.
+        if (isa<memref::SubViewOp>(user)) return;
+        if (group.liveness.count(user)) {
+          aliasGroups.push_back(i);
+          break;
+        }
+      }
+    }
+    if (aliasGroups.empty()) {
+      // If we didn't find any alias group create a new one.
+      AllocGroup &newGroup = groups.emplace_back();
+      newGroup.allocs.push_back(alloc);
+      Liveness::OperationListT liveInfo =
```

Comment:
Oh what!! The MLIRs liveness data structure is also a list of all operations. See the big comment there

```
/// Gets liveness info (if any) for the given value.
/// This includes all operations in which the given value is live.
/// Note that the operations in this list are not ordered and the current
/// implementation is computationally expensive (as it iterates over all
/// blocks in which the given value is live).
```

These are kind of things that can blow up compilation time (granted we only have a few instructions in the codegen dispatch hopefully)

Reply:
Yes, this is why the liveness is DenseSet of Operation. Since we wouldn't have more Operations than the total number of operations that shouldn't explode but this is why I need to represent the alloc structure the way they are.